File: projects/antia_numerical_methods@Beliavsky/f77/sub16/byn.f
!	To calculate the Bessel function of second kind of integral order
!		for real argument
!
!	N : (input) Order of Bessel function required, N may be negative
!		or positive. For N=0,1 use BY0 and BY1 respectively.
!	X : (input) Argument at which the value is required
!	BY : (output) Real array of length at least ABS(N)+1
!		which will contain the value of Bessel function of order
!		0,1,...,ABS(N). BY(I+1) will contain Bessel function of order I
!		or -I (if N<0).
!
!	Required routines : BY0, BY1
!
 
      SUBROUTINE BYN(N,X,BY)
      IMPLICIT REAL*16(A-H,O-Z)
      DIMENSION BY(N+1)
 
      BY(1)=BY0(X)
      BY(2)=BY1(X)
      IF(X.LE.0.0) RETURN
!	Use the recurrence relation in Forward direction
      DO 2000 I=2,ABS(N)
        BY(I+1)=2*(I-1)*BY(I)/X-BY(I-1)
2000  CONTINUE
 
      IF(N.LT.0) THEN
        DO 2500 I=2,ABS(N)+1,2
          BY(I)=-BY(I)
2500    CONTINUE
      ENDIF
      END

File: projects/antia_numerical_methods@Beliavsky/c/fun/byn.c
/*	To calculate the Bessel function of second kind of integral order
		for real argument

	N : (input) Order of Bessel function required, N may be negative
		or positive. For N=0,1 use BY0 and BY1 respectively.
	X : (input) Argument at which the value is required
	BY : (output) Array of length at least abs(N)+1
		which will contain the value of Bessel function of order
		0,1,...,ABS(N). BY[I] will contain Bessel function of order I
		or -I (if N<0).

	Required functions : BY0, BY1
*/

#include <math.h>

double by0(double x);
double by1(double x);

void byn(int n, double x, double by[])

{
	int i;
 
	by[0]=by0(x);
	by[1]=by1(x);
	if(x<= 0.0) return;
/*	Use the recurrence relation in Forward direction */
	for(i=2; i<=abs(n); ++i) by[i]=2*(i-1)*by[i-1]/x-by[i-2];

	if(n<0) {
		for(i=1; i<=abs(n); i +=2) by[i]= -by[i];
	}
	return;
}
